### PR TITLE
Fix retain cycle between cell and view controller

### PR DESCRIPTION
**Pull request by DreamingInBinary**
In reference to [this issue](https://github.com/TextureGroup/Texture/issues/1720#issuecomment-549203953).

### CODE REVIEW

**Comment by bolsinga** [comment]
Please add a unit test that verifies the cycle is gone. See #1429 for an example. Thanks!

**Comment by rahul-malik** [comment]
@dreaminginbinary Mind updating the PR with a test and signing the CLA so we can merge this?

**Comment by DreamingInBinary** [comment]
@rahul-malik Sorry about the radio silence - had a few things come up and went out for the holidays. I'll get this in this week 👍

**Comment by DreamingInBinary** [comment]
@rahul-malik tests added 👍

File: Podfile.lock [hunk 1]
```diff
@@ -46,10 +27,7 @@ SPEC CHECKSUMS:
...
   OCMock: 2cd0716969bab32a2283ff3a46fd26a8c8b4c5e3
-  PINCache: 7cb9ae068c8f655717f7c644ef1dff9fd573e979
-  PINOperation: a6219e6fc9db9c269eb7a7b871ac193bcf400aac
-  PINRemoteImage: 81bbff853acc71c6de9e106e9e489a791b8bbb08
 
-PODFILE CHECKSUM: 445046ac151568c694ff286684322273f0b597d6
+PODFILE CHECKSUM: 345a6700f5fdec438ef5553e1eebf62653862733
 
-COCOAPODS: 1.6.0
+COCOAPODS: 1.8.4
```
Comment (bolsinga) [inline Podfile.lock:33]: I wouldn't expect to see project nor podfile changes with this diff.

**Comment by DreamingInBinary** [comment]
@bolsinga I wasn't sure of the best way to undo the `podfile.lock` changes, so I updated from the lock file from Master. Typically if I have issues with a lock file, I delete it and run `pod install` again but I wasn't sure if removing the file would cause any other issues 🤔

**Comment by bolsinga** [comment]
> @bolsinga I wasn't sure of the best way to undo the `podfile.lock` changes, so I updated from the lock file from Master. Typically if I have issues with a lock file, I delete it and run `pod install` again but I wasn't sure if removing the file would cause any other issues 🤔

I think you need to merge with master...

File: AsyncDisplayKit.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -1390,8 +1392,6 @@
 				83A7D95D1D446A6E00BF333E /* ASWeakMapTests.mm */,
 				CC3B208D1C3F7D0A00798563 /* ASWeakSetTests.mm */,
 				695BE2541DC1245C008E6EA5 /* ASWrapperSpecSnapshotTests.mm */,
-				407B8BAD2310E2ED00CB979E /* ASLayoutSpecUtilitiesTests.mm */,
-				CC698826247855F200487428 /* UIImage+ASConvenienceTests.mm */,
```
[the remark this answers is not in the document]
Comment (DreamingInBinary) [reply AsyncDisplayKit.xcodeproj/project.pbxproj:1394]: @bolsinga Sorry for the back and forth here Greg! 

So I took master's .pbxproj, and then added back in my new file (`ASCellNodeTests.mm`)  and it doesn't appear to remove `ASLayoutSpecUtilitiesTests.mm` or `UIImage+ASConvenienceTests.mm` but rather seems to shift where they are in that particular file if I'm seeing things correctly.

[the remark this answers is not in the document]
Comment (bolsinga) [reply AsyncDisplayKit.xcodeproj/project.pbxproj:1394]: Thanks! Sorry for my delay in finding your response and update.

**Comment by bolsinga** [comment]
Looks like the license still needs to be accepted. I’ll land once that is complete. Thanks!

**Comment by DreamingInBinary** [comment]
@bolsinga Done!
Screen Shot 2020-06-18 at 9 51 50 AM

**Comment by bolsinga** [comment]
> @bolsinga Done!
> Screen Shot 2020-06-18 at 9 51 50 AM

In the UI I see, it still says "Jordan Morgan" has not signed this, but "DreamingInBinary" has. Maybe those two accounts are not fully associated with each other? I don't know why there are 2 associated with this PR in the first place...

**Comment by mycroftcanner** [comment]
@bolsinga did you manage to sign?

**Comment by bolsinga** [comment]
> @bolsinga did you manage to sign?

Hi @mycroftcanner It's @DreamingInBinary that needs to sign the CLA. @DreamingInBinary would you be able to remove the unsigned committer, or sign them? Thanks!

**Comment by foxware00** [comment]
@bolsinga @DreamingInBinary Interesting in the PR.

I've got an ASPagerNode setup and as I move out of range of a pager it's being retained indefinitely. As I'm using the AsPagerNode as one of my main tabs, this memory never gets freed.

I've used the fix applied in the PR. However, there seem to be unwanted side effects.

The ASPagerNode and I believe all ASCollectionNode's initialize their dataset and expect all nodes to be left around. The view controller included. The problem is, this now (I think correctly, deinit's each view controller as it goes off screen). The problem is the ASPagerNode never re-creates it so you end up with empty pages. See the ASPagerNode documentation below. It seems to indicate that it's not called when the nodes are about to display and seems to be called in one batch as the dataset is applied. This means, once deallocated pages, they're not able to be re-created.

I'm curious on two fronts, this fix in the PR seems correct to me. Nodes not in the range, that have disappeared should be allowed to deinit and free up memory. They should then be recreated. Neither the current logic or the PR allow that. Maybe a tweak is needed in re-creating nodes as they come into view. 

@nguyenhuy @rahul-malik might be able to shed some light, but it looks like this PR has unwanted side effects or at the very least is a breaking change based on the assumptions of ASCollectionNode's allocation. I'd love to understand this a little more.

```
**
 * This method replaces -collectionView:nodeForItemAtIndexPath:
 *
 * @param pagerNode The sender.
 * @param index     The index of the requested node.
 * @return a node for display at this index. This will be called on the main thread and should
 *   not implement reuse (it will be called once per row).  Unlike UICollectionView's version,
 *   this method is not called when the row is about to display.
 */
- (ASCellNode *)pagerNode:(ASPagerNode *)pagerNode nodeAtIndex:(NSInteger)index;
```

> As mentioned above, since ASCellNodes are not meant to be reused, they have a longer lifecycle compared to the view or layer that they encapsulate or their corresponding UICollectionViewCell or UITableViewCell. ASCellNodes are deallocated when they are no longer used and removed from the container node. This can occur after a batch update that includes a reload data or deletion, or after the container node is no longer used and thus released.

>Furthermore, the current implementations of ASCollectionNode and ASTableNode allocate all cell nodes as soon as they are inserted. That is, if clients perform a batch update that inserts 100 items into a collection node, the collection node will allocate 100 cell nodes as part of the processing of the batch update. It will also perform a layout calculation on each of the new cell nodes. Thus, at the end of the process, the collection node will manage 100 nodes more and these nodes have calculated layouts ready to be used.

From [here](https://texturegroup.org/development/node-lifecycle.html)

**Pull request closed without merging**